projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx

```diff
@@ -0,0 +1,119 @@
+import React, { useCallback, useEffect, useMemo, useState } from "react";
+import { TokenInput } from "src/components/Swap/TokenInput";
+import { Token, TokenValue } from "@beanstalk/sdk";
+import { useAllTokensBalance } from "src/tokens/useTokenBalance";
+import styled from "styled-components";
+import { images } from "src/assets/images/tokens";
+import { useAccount } from "wagmi";
+import { ContractReceipt } from "ethers";
+import { Well } from "@beanstalk/sdk/Wells";
+import { useQuery } from "@tanstack/react-query";
+
+type LiquidityAmounts = {
+  [key: number]: TokenValue;
+};
+
+type AddLiquidityProps = {
+  well: Well;
+};
+
+export const AddLiquidity = ({ well }: AddLiquidityProps) => {
+  const { address } = useAccount();
+  const [wellTokens, setWellTokens] = useState<Token[]>([]);
+  const [amounts, setAmounts] = useState<LiquidityAmounts>({});
+  const [receipt, setReceipt] = useState<ContractReceipt | null>(null);
+  const [isLoadingAllBalances, setIsLoadingAllBalances] = useState(true);
+
+  const { isLoading: isAllTokenLoading, refetch: refetchBalances } = useAllTokensBalance();
+
+  const atLeastOneAmountNonzero = () => Object.values(amounts).filter((amount) => amount.value.gt("0")).length > 0;
+
+  useEffect(() => {
+    const fetching = isAllTokenLoading;
+    fetching ? setIsLoadingAllBalances(true) : setTimeout(() => setIsLoadingAllBalances(false), 500);
+  }, [isAllTokenLoading]);
+
+  useEffect(() => {
+    if (well.tokens) {
+      const tokens: Token[] = [];
+      const initialAmounts: LiquidityAmounts = {};
+      well.tokens.forEach((token, index) => {
+        token.setMetadata({ logo: images[token.symbol] ?? images.DEFAULT });
+        tokens.push(token);
+        initialAmounts[index] = TokenValue.ZERO;
+      });
+
+      setWellTokens(tokens);
+      setAmounts(initialAmounts);
+    }
+  }, [well]);
+
+  const {
+    data: quote,
+    isLoading: loadingQuote,
+    isError: quoteError
+  } = useQuery(["wells", address, amounts], async () => {
+    if (!atLeastOneAmountNonzero()) {
+      return null;
+    }
+    return well.addLiquidityQuote(Object.values(amounts));
+  });
+
+  const addLiquidityButtonClickHandler = useCallback(async () => {
+    if (quote && address) {
+      const addLiquidityTxn = await well.addLiquidity(Object.values(amounts), quote, address);
+      const receipt = await addLiquidityTxn.wait();
+      setReceipt(receipt);
+      refetchBalances();
+    }
+  }, [well.addLiquidity, amounts, quote, address]);
+
+  const handleInputChange = useCallback(
+    (index: number) => (a: TokenValue) => {
+      setAmounts({ ...amounts, [index]: a });
+    },
+    [amounts]
+  );
+
+  const addLiquidityButtonEnabled = address && atLeastOneAmountNonzero();
```

Comment:
can cache this with useMemo so it's not eval'ed every time
```suggestion
  const addLiquidityButtonEnabled = useMemo(() => address && atLeastOneAmountNonzero(), [addres, atLeastOneAmountNonzero()]);
```

Reply:
Will apply in the next PR.